### ISSUE TITLE
return negative value from from_bytes

### ISSUE THREAD

**Issue by User-Man** [etingof/pyasn1#84]
Hi guys, I am new in ASN.1 coding and try handle snmp trap and have some problems with some equipment

Most of trap working good, but some crash with error:

```
 Traceback (most recent call last):
  File "./broken_test.py", line 17, in <module>
    a, b = pyasn_decoder.decode(whole_msg, asn1Spec = proto.Message())
  File "/root/pyasn1/codec/ber/decoder.py", line 1160, in __call__
    **options
  File "/root/pyasn1/codec/ber/decoder.py", line 496, in valueDecoder
    component, head = decodeFun(head, asn1Spec, **options)
  File "/root/pyasn1/codec/ber/decoder.py", line 1160, in __call__
    **options
  File "/root/pyasn1/codec/ber/decoder.py", line 686, in valueDecoder
    tagSet, length, state, **options
  File "/root/pyasn1/codec/ber/decoder.py", line 1160, in __call__
    **options
  File "/root/pyasn1/codec/ber/decoder.py", line 496, in valueDecoder
    component, head = decodeFun(head, asn1Spec, **options)
  File "/root/pyasn1/codec/ber/decoder.py", line 1160, in __call__
    **options
  File "/root/pyasn1/codec/ber/decoder.py", line 523, in valueDecoder
    component, head = decodeFun(head, asn1Spec, **options)
  File "/root/pyasn1/codec/ber/decoder.py", line 1160, in __call__
    **options
  File "/root/pyasn1/codec/ber/decoder.py", line 496, in valueDecoder
    component, head = decodeFun(head, asn1Spec, **options)
  File "/root/pyasn1/codec/ber/decoder.py", line 1160, in __call__
    **options
  File "/root/pyasn1/codec/ber/decoder.py", line 686, in valueDecoder
    tagSet, length, state, **options
  File "/root/pyasn1/codec/ber/decoder.py", line 1160, in __call__
    **options
  File "/root/pyasn1/codec/ber/decoder.py", line 686, in valueDecoder
    tagSet, length, state, **options
  File "/root/pyasn1/codec/ber/decoder.py", line 1160, in __call__
    **options
  File "/root/pyasn1/codec/ber/decoder.py", line 686, in valueDecoder
    tagSet, length, state, **options
  File "/root/pyasn1/codec/ber/decoder.py", line 1160, in __call__
    **options
  File "/root/pyasn1/codec/ber/decoder.py", line 109, in valueDecoder
    return self._createComponent(asn1Spec, tagSet, value), tail
  File "/root/pyasn1/codec/ber/decoder.py", line 45, in _createComponent
    return asn1Spec.clone(value)
  File "/root/pyasn1/type/univ.py", line 273, in clone
    return base.AbstractSimpleAsn1Item.clone(self, value, **kwargs)
  File "/root/pyasn1/type/base.py", line 311, in clone
    return self.__class__(value, **initilaizers)
  File "/root/pyasn1/type/univ.py", line 72, in __init__
    base.AbstractSimpleAsn1Item.__init__(self, value, **kwargs)
  File "/root/pyasn1/type/base.py", line 209, in __init__
    raise exType('%s at %s' % (exValue, self.__class__.__name__))
pyasn1.type.error.ValueConstraintError: ConstraintsIntersection(ValueRangeConstraint(0, 4294967295)) failed at: ValueConstraintError('ValueRangeConstraint(0, 4294967295) failed at: ValueConstraintError(-112878092,)',) at TimeTicks
```
Reading code I find in pyasn1-0.3.7-py2.7.egg/pyasn1/compat/integer.py

```python
29   if signed and oct2int(octets[0]) & 0x80:
30       return value - (1 << len(octets) * 8)
```
Can you explain me please, what means "oct2int(octets[0]) & 0x80"?
When me code crash: oct2int(octets[0]) = 249

Thanks!

**Comment by etingof** [comment]
TL;DR; You seem to receive a TRAP message with an integer value (-112878092 ) which is out of defined range (0, 4294967295).

To answer your question: the SNMP message is serialized into a stream of octets. On reception, the octets get interpreted (thus `oct2int()`) as various data types, including integers. When negative integers are serialized, the minus sign is indicated in encoding by the very leading bit of the first octet set to 1. Thus you have `0x80` there to test that.

You may want to enable pyasn1 debugging to get more details on the failing field:

    from pyasn1 import debug
    debug.setLogger(debug.Debug('all'))

One more possibly [relevant link](http://pysnmp.sourceforge.net/faq/ignored-snmp-packets.html).

**Comment by User-Man** [comment]
Thank you, I will by try.
